modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistentcut/AbstractConsistentCutTest.java

```diff
@@ -0,0 +1,301 @@
+
+        startGrids(nodes());
+
+        grid(0).cluster().state(ClusterState.ACTIVE);
+
+        startClientGrid(nodes());
+
+        snp(grid(0)).createSnapshot(SNP).get();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
+
+    /**
+     * @return {@code true} if NearCache is enabled, otherwise {@code false}.
+     */
+    protected boolean withNearCache() {
+        return false;
+    }
+
+    /**
+     * @return Number of server nodes.
+     */
+    protected abstract int nodes();
+
+    /**
+     * @return Number of backups for cache.
+     */
+    protected abstract int backups();
+
+    /** */
+    protected void awaitSnapshotResourcesCleaned() {
+        try {
+            GridTestUtils.waitForCondition(() -> {
+                boolean ready = true;
+
+                for (Ignite g: G.allGrids())
+                    ready &= snp((IgniteEx)g).currentCreateRequest() == null;
+
+                return ready;
+            }, 10);
+        }
+        catch (IgniteInterruptedCheckedException e) {
+            throw new IgniteException(e);
+        }
+    }
+
+    /**
+     * Checks WALs for correct Consistency Cut.
+     *
+     * @param txCnt Count of run transactions.
+     * @param cuts    Number of Consistent Cuts was run within a test.
+     */
+    protected void checkWalsConsistency(int txCnt, int cuts) throws Exception {
+        List<ConsistentCutWalReader> readers = new ArrayList<>();
+
+        for (int i = 0; i < nodes(); i++) {
+            ConsistentCutWalReader reader = new ConsistentCutWalReader(walIter(i));
+
+            readers.add(reader);
+
+            reader.read();
+
+            int expCuts = reader.cuts.get(reader.cuts.size() - 1).id != null ? cuts : cuts + 1;
+
+            assertEquals(expCuts, reader.cuts.size());
+        }
+
+        // Transaction ID -> (cutId, nodeId).
+        Map<GridCacheVersion, T2<UUID, Integer>> txMap = new HashMap<>();
+
+        // Includes incomplete state also.
+        for (int cutId = 0; cutId < cuts + 1; cutId++) {
+            for (int nodeId = 0; nodeId < nodes(); nodeId++) {
+                // Skip if the latest cut is completed.
+                if (readers.get(nodeId).cuts.size() == cutId)
+                    continue;
+
+                ReadConsistentCut cut = readers.get(nodeId).cuts.get(cutId);
+
+                for (GridCacheVersion xid: cut.txs) {
+                    T2<UUID, Integer> prev = txMap.put(xid, new T2<>(cut.id, nodeId));
+
+                    if (prev != null) {
+                        assertTrue("Transaction miscutted: [xid=" + xid + ", node" + prev.get2() + "=" + prev.get1() +
+                            ", node" + nodeId + "=" + cut.id,
+                            Objects.equals(prev == null ? null : prev.get1(), cut.id));
+                    }
+                }
+            }
+        }
+
+        assertEquals(txCnt, txMap.size());
```

Comment:
It seems we need more complex check of transactions here.
We need to check that xid presented in all local snapshots where we expect it to be.

Reply:
see AbstractConsistentCutTest#checkWalsConsistency